### PR TITLE
Generate provenance files

### PR DESCRIPTION
**Pull request by rbehjati**
Updates the `build` command line tool to generate provenance files if the build is successful.

### CODE REVIEW

File: common/common.go [hunk 1]
```diff
@@ -243,26 +254,74 @@ func (b *BuildConfig) VerifyBinarySha256Hash() error {
...
+		Type:          InTotoStatementV01,
+		Subject:       []slsa.Subject{subject},
+		PredicateType: SLSAPredicateV02,
+		Predicate:     predicate,
+	}, nil
+}
+
+func parseBuilderImageURI(imageURI string) (string, string, error) {
+	// We expect the URI of the builder image to be of the form NAME@DIGEST
+	uRIParts := strings.Split(imageURI, "@")
```
Comment (tiziano88) [inline common/common.go:314]:
```suggestion
	URIParts := strings.Split(imageURI, "@")
```

**Review by tiziano88 — APPROVED** [review]

File: cmd/build/main.go [hunk 2]
```diff
@@ -27,9 +29,23 @@ func main() {
 		"Required - Path to a toml file containing the build configs.")
 	gitRootDirPtr := flag.String("git_root_dir", "",
 		"Optional - Root of the Git repository. If not specified, sources are fetched from the repo specified in the config file.")
+	provenancePathPtr := flag.String("provenance_path", "",
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply cmd/build/main.go]: Thanks. Fixed.

File: common/common.go (hunk 1 above)
Comment (rbehjati) [reply common/common.go]: Done.

File: common/common.go [hunk 3]
```diff
@@ -243,26 +254,74 @@ func (b *BuildConfig) VerifyBinarySha256Hash() error {
...
 
 	if b.ExpectedBinarySha256Hash != "" && b.ExpectedBinarySha256Hash != binarySha256Hash {
-		return fmt.Errorf("the hash of the output binary does not match the expected binary hash; got %s, want %v",
+		return nil, fmt.Errorf("the hash of the output binary does not match the expected binary hash; got %s, want %v",
 			binarySha256Hash, b.ExpectedBinarySha256Hash)
 	}
 
-	return nil
+	subject := slsa.Subject{
+		// TODO: Get the name as an input in the TOML file.
```
Comment (mariaschett) [inline common/common.go:274]: Do we want to open a issue here?

File: common/common.go [hunk 4]
```diff
@@ -243,26 +254,74 @@ func (b *BuildConfig) VerifyBinarySha256Hash() error {
...
+	return &slsa.Provenance{
+		Type:          InTotoStatementV01,
+		Subject:       []slsa.Subject{subject},
+		PredicateType: SLSAPredicateV02,
+		Predicate:     predicate,
+	}, nil
+}
+
+func parseBuilderImageURI(imageURI string) (string, string, error) {
+	// We expect the URI of the builder image to be of the form NAME@DIGEST
```
Comment (mariaschett) [inline common/common.go:313]: nit: Maybem move this as a toplevel comment to to the function?

File: common/common_test.go [hunk 5]
```diff
@@ -67,6 +67,23 @@ func TestLoadBuildConfigFromProvenance(t *testing.T) {
 	checkBuildConfig(config, t)
 }
 
+func TestParseBuilderImageURI(t *testing.T) {
```
Comment (mariaschett) [inline common/common_test.go:70]: Nice! :) 

I was thinking if a test failing to parse an imageURI makes sense, too, but I am not too convinced myself.

File: cmd/build/main.go [hunk 6]
```diff
@@ -16,20 +16,36 @@
...
 	"log"
+	"os"
 
 	"github.com/project-oak/transparent-release/build"
 )
 
 func main() {
-	buildConfigPathPtr := flag.String("config", "",
+	buildConfigPath := flag.String("build_config_path", "",
 		"Required - Path to a toml file containing the build configs.")
```
Comment (mariaschett) [inline cmd/build/main.go:29]: nit & pre-existing: "a" toml file vs -> "the" toml file --- would we expect there to be more than one toml file in general? Maybe also want to clarify that these are the build configs to build the gitRootDir? Is that correct?

File: cmd/build/main.go [hunk 7]
```diff
@@ -16,20 +16,36 @@
...
-	gitRootDirPtr := flag.String("git_root_dir", "",
+	gitRootDir := flag.String("git_root_dir", "",
 		"Optional - Root of the Git repository. If not specified, sources are fetched from the repo specified in the config file.")
```
Comment (mariaschett) [inline cmd/build/main.go:31]: nit: `config file` -> `build config file`? Or not?

**Review by mariaschett — APPROVED** [review]
Nice :)

File: common/common.go (hunk 4 above)
Comment (rbehjati) [reply common/common.go:313]: The function is not exposed, so it does not really require documentation. I think this comment is clearer here.

File: common/common_test.go (hunk 5 above)
Comment (rbehjati) [reply common/common_test.go]: Good point. Added a negative test.

**Review by rbehjati — COMMENTED** [review]
Thanks for the review.

**Pull request merged**